platooning_strategic_IHP/src/platoon_strategic_ihp.cpp

```diff
@@ -1168,14 +1615,95 @@ namespace platoon_strategic_ihp
             return MobilityRequestResponse::NO_RESPONSE;
         }
     }
+    
+    // UCLA: add request call-back function for lead with operation state (for cut-in join)
+    MobilityRequestResponse PlatoonStrategicIHPPlugin::mob_req_cb_leadwithoperation(const cav_msgs::MobilityRequest& msg)
+    {
+        /*
+        *   Current leader leading while opening the gap for cut-in join.
+        */
+
+        // Check request plan type  
+        cav_msgs::PlanType plan_type = msg.plan_type;
+        if (plan_type.type == cav_msgs::PlanType::PLATOON_CUTIN_JOIN) 
+        {
+            // task 1. slow down to create gap 
+            int join_index = msg.join_index;
+            // determine if joining vehicle in position
+            double cutinDtdDifference = current_downtrack_ - pm_.platoon[0].vehiclePosition - config_.vehicleLength;
+            bool isFrontJoinerInPosition = (cutinDtdDifference <= 1.5*config_.vehicleLength) || (cutinDtdDifference >= -1.5*config_.vehicleLength);
+            // todo: reserved for cut-in from middle/rear
+            // bool isMidJoinerInPosition = ((pm_.platoon[join_index].vehiclePosition - current_downtrack_ <= 3*config_.vehicleLength)); 
+            // bool isRearJoinerInPosition = (pm_.platoon[pm_.platoon.size()-1].vehiclePosition - current_downtrack_ <= 3*config_.vehicleLength);
+
+            if (join_index == -1 && isFrontJoinerInPosition )
+            {
+                // slow down leader to allow joiner
```

Comment:
If joiner is farther away than 1 vehicle length, then host should not have to slow down and create a gap.